{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The watermark extension is already loaded. To reload it, use:\n",
      "  %reload_ext watermark\n",
      "Author: Filippo_Valle\n",
      "\n",
      "Last updated: Wed May 05 2021\n",
      "\n",
      "Python implementation: CPython\n",
      "Python version       : 3.8.8\n",
      "IPython version      : 7.22.0\n",
      "\n",
      "pandas    : 1.2.4\n",
      "numpy     : 1.19.0\n",
      "scipy     : 1.6.2\n",
      "matplotlib: 3.4.1\n",
      "seaborn   : 0.11.1\n",
      "\n",
      "Compiler    : GCC 9.3.0\n",
      "OS          : Linux\n",
      "Release     : 5.11.0-16-generic\n",
      "Machine     : x86_64\n",
      "Processor   : x86_64\n",
      "CPU cores   : 12\n",
      "Architecture: 64bit\n",
      "\n",
      "Git hash: bb46305b8fc98e761c0b4bea818075dc7f7ecb6d\n",
      "\n",
      "Git repo: git@github.com:SilviaLazzardi/The_single_cell_transcriptome_as_a_component_system\n",
      "\n",
      "Git branch: main\n",
      "\n",
      "Watermark: 2.2.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m  -u -n -p pandas,numpy,scipy,matplotlib,seaborn -a Filippo_Valle -g -r -b -w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "import os, sys, gc, pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "\n",
    "from frontiers_colors import get_color\n",
    "from frontiers_analysis import heaps, load_tissue, save_model, mazzolini, load_all_data\n",
    "\n",
    "from tableanalyser import plotvarmen, plotcv2mean\n",
    "from tacos_plot import scatterdense"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Choose one of these data source:\n",
    "* mca for Mouse Cell Atlas (default)\n",
    "* tm for Tabula Muris\n",
    "* gtex for GTex\n",
    "\n",
    "For Mouse Cell Atlas data will be automatically downloaded!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_source = \"mca\"\n",
    "os.system(f\"mkdir -p {data_source}\")\n",
    "files = os.listdir(data_source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download all the files for MCA\n",
    "from frontiers_analysis import get_files\n",
    "get_files()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please download the following files from original dataset\n",
    "[https://figshare.com/s/865e694ad06d5857db4b](https://figshare.com/s/865e694ad06d5857db4b):\n",
    "* *MCA_Figure2_Cell.Info.xlsx*\n",
    "* *MCA_CellAssignments.csv*\n",
    "\n",
    "They are required before run the following cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.read_excel(\"MCA_Figure2_Cell.Info.xlsx\", index_col=0)\n",
    "\n",
    "df_info['Tissue'] = list(map(lambda ann: ann.replace(\"Adult\",\"\"), df_info['Tissue'])) #remove adult or fetal\n",
    "df_info['Tissue'] = list(map(lambda ann: ann.replace(\"Fetal\",\"\"), df_info['Tissue'])) #remove adult or fetal\n",
    "df_info['Tissue'] = list(map(lambda ann: ann.replace(\"Neonatal\",\"\"), df_info['Tissue'])) #remove adult or fetal\n",
    "df_info.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_file=pd.read_csv(\"MCA_CellAssignments.csv\", index_col=0).set_index('Cell.name')\n",
    "df_file['Cell_type'] = list(map(lambda ann: ann.split('(')[0].split(' ')[0], df_file['Annotation'])) #remove adult or fetal\n",
    "df_file.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = df_file\n",
    "df_info[\"id\"] = df_file.index.values\n",
    "df_info.rename(columns={'Tissue':'tissue',\"id\": \"Cell_id\"}, inplace=True)\n",
    "df_info.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = list(filter(lambda f: \"dge\" in f,  os.listdir(data_source)))\n",
    "files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyse organs and store data\n",
    "Here some files with all the analyses will be stored, this take a long time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(files[0])\n",
    "\n",
    "saved = []\n",
    "frac_of = {}\n",
    "\n",
    "if \"mca\" in data_source:\n",
    "    data = pd.read_csv(f\"{data_source}/{files[0]}\", sep=\" \")\n",
    "    data = data[data.columns[data.columns.isin(df_info['Cell_id'])]]\n",
    "\n",
    "    #tissues = df_info[df_info['Cell_id'].isin(data.columns) & (df_info['Development_stage']=='Adult')]['tissue'].unique() # only adults\n",
    "    tissues = df_info[df_info['Cell_id'].isin(data.columns)]['tissue'].unique()\n",
    "\n",
    "if \"tm\" in data_source:\n",
    "    tissues = [files[0]]\n",
    "\n",
    "\n",
    "print(tissues)\n",
    "for tissue in tissues:\n",
    "    if \"mca\" in data_source:\n",
    "        df = data.reindex(columns=df_info[df_info['Cell_id'].isin(data.columns) & (df_info['tissue']==tissue)]['Cell_id'])\n",
    "        #df = df.divide(df.sum(0), 1).applymap(lambda cpm: np.log(cpm*1e6+1)) #log(cpm +1)\n",
    "    \n",
    "    if \"tm\" in data_source:\n",
    "        df = pd.read_csv(f\"{data_source}/{files[0]}\", sep=\",\", index_col=0) #tabula muris\n",
    "        df = clean_df(df) #tabula muris\n",
    "    \n",
    "    A = df.sum(axis=1)\n",
    "    O = df.apply(lambda x: len(x[x>0]), 1)\n",
    "    M = df.apply(np.sum, 0)\n",
    "    global_f = df.divide(M, axis=1).sum(1)\n",
    "    cell_zeros = df.astype(int).apply(lambda x: len(x[x==0]), 0)\n",
    "    gene_presence = df.apply(lambda x: len(x[~x.isna()]), 1).astype(int)\n",
    "    gene_presence_nonnull = df.apply(lambda x: len(x[~x.isna() & (x>0)]), 1).astype(int)\n",
    "    var = df.apply(lambda x: (x*x).sum(), 1)\n",
    "    \n",
    "    frac_of[tissue] = {'data': A,\n",
    "                            'N': len(M)}\n",
    "    \n",
    "    diffWords = df.apply(lambda x: len(x[x>0]), 0)\n",
    "    n_genes = {}\n",
    "    n_genes[tissue] = diffWords\n",
    "    \n",
    "    n_expressed_genes = pd.Series(index=df.index, data=np.zeros_like(df.index))\n",
    "    n_expressed_genes[A.sort_values(ascending=False).index[:100]]+=1\n",
    "    save_model(df, \"data\", tissue)\n",
    "    mazzolini(M, A/A.sum(), tissue)\n",
    "    #heaps(M, diffWords, tissue)\n",
    "    \n",
    "    saved.append(tissue)\n",
    "    df.to_csv(f\"{data_source}/mainTable_{tissue}.csv\", index=True, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in files[1:]:\n",
    "    print(file)\n",
    "    if \"mca\" in data_source:\n",
    "        data = pd.read_csv(f\"{data_source}/{file}\", sep=\" \")\n",
    "        data = data[data.columns[data.columns.isin(df_info['Cell_id'])]]\n",
    "\n",
    "        if len(data.columns) < 1:\n",
    "            print(f\"skipping {file} \\n\")\n",
    "            continue\n",
    "        #tissues = df_info[df_info['Cell_id'].isin(data.columns) & (df_info['Development_stage']=='Adult')]['tissue'].unique() # only adults\n",
    "        tissues = df_info[df_info['Cell_id'].isin(data.columns)]['tissue'].unique()\n",
    "\n",
    "    if \"tm\" in data_source:\n",
    "        data = pd.read_csv(f\"{data_source}/{file}\", sep=\",\", index_col=0)\n",
    "        tissues = [files[0]]\n",
    "    \n",
    "    print(tissues)\n",
    "    if len(tissues) < 1:\n",
    "        print(f\"no tissue found \\n\")\n",
    "        continue\n",
    "    for tissue in tissues:\n",
    "        \n",
    "        if \"mca\" in data_source:\n",
    "            df = data.reindex(columns=df_info[df_info['Cell_id'].isin(data.columns) & (df_info['tissue']==tissue)]['Cell_id'])\n",
    "        \n",
    "        if \"tm\" in data_source:\n",
    "            df = clean_df(data)\n",
    "        \n",
    "        if tissue in saved:\n",
    "            df = df.transpose().append(pd.read_csv(f\"{data_source}/mainTable_{tissue}.csv\", index_col=0, header=0).transpose(), ignore_index=False, sort=True).transpose().fillna(0)\n",
    "        \n",
    "        #df = df.divide(df.sum(0), 1).applymap(lambda cpm: np.log(cpm*1e6+1)) #log(cpm +1)\n",
    "        A_sub = df.sum(axis=1)\n",
    "        O_sub = df.apply(lambda x: len(x[x>0]), 1)\n",
    "        M_sub = df.apply(np.sum, 0)\n",
    "        cell_zeros_sub = df.apply(lambda x: len(x[x>0]), 0)\n",
    "        diffWords_sub = df.apply(lambda x: len(x[x>0]), 0)\n",
    "        \n",
    "        if tissue in frac_of.keys():\n",
    "            frac_of[tissue]['data'].add(A_sub, fill_value = 0)\n",
    "            frac_of[tissue]['N']+=len(M_sub)\n",
    "        else:\n",
    "            frac_of[tissue] = {'data': A_sub,\n",
    "                                  'N': len(M_sub)}\n",
    "\n",
    "        A = A.add(A_sub, fill_value=0)\n",
    "        O = O.add(O_sub, fill_value=0)\n",
    "        M = M.append(M_sub)\n",
    "        global_f.add(df.divide(M_sub, axis=1).sum(1))\n",
    "        cell_zeros = cell_zeros.append(cell_zeros_sub)\n",
    "        diffWords = diffWords.append(diffWords_sub)\n",
    "        if tissue in n_genes.keys():\n",
    "            n_genes[tissue] = n_genes[tissue].append(diffWords_sub)\n",
    "        else:\n",
    "            n_genes[tissue]= diffWords_sub\n",
    "        \n",
    "        gene_presence = gene_presence.add(df.apply(lambda x: len(x[~x.isna()]), 1), fill_value=0)\n",
    "        gene_presence_nonnull_sub = df.apply(lambda x: len(x[~x.isna() & (x>0)]), 1).astype(int)\n",
    "        gene_presence_nonnull = gene_presence_nonnull.add(gene_presence_nonnull_sub, fill_value=0).astype(int)\n",
    "        var = var.add(df.apply(lambda x: (x*x).sum(), 1), fill_value=0)\n",
    "        \n",
    "        n_expressed_genes_sub = pd.Series(index=df.index, data=np.zeros_like(df.index))\n",
    "        n_expressed_genes_sub[A_sub.sort_values(ascending=False).index[:100]]+=1\n",
    "        n_expressed_genes = n_expressed_genes.add(n_expressed_genes_sub, fill_value=0)\n",
    "\n",
    "        save_model(df,\"data\",tissue)\n",
    "        mazzolini(M_sub, A_sub/A_sub.sum(), tissue)\n",
    "        #heaps(M_sub, diffWords_sub, tissue)\n",
    "        saved.append(tissue)\n",
    "        df.to_csv(f\"{data_source}/mainTable_{tissue}.csv\", index=True, header=True)\n",
    "        print(f\"Handling {len(M)} cells\")\n",
    "\n",
    "        del df \n",
    "        del A_sub\n",
    "        del O_sub\n",
    "        del M_sub\n",
    "        del diffWords_sub\n",
    "        del gene_presence_nonnull_sub\n",
    "        gc.collect()\n",
    "        print(\"\\n\")\n",
    "    del tissues\n",
    "    del data\n",
    "    gc.collect()\n",
    "\n",
    "means = A/gene_presence\n",
    "means_nozero = A/gene_presence_nonnull\n",
    "var = var/gene_presence - means*means\n",
    "f = (A/A.sum())\n",
    "O = O / gene_presence\n",
    "O = O.reindex_like(means)\n",
    "cv2 = var/means/means\n",
    "\n",
    "\n",
    "print(\"Saving data.pkl\")\n",
    "data = {\n",
    "    'means': means,\n",
    "    'var': var,\n",
    "    'freq': A/A.sum(),\n",
    "    'O': O,\n",
    "    'M': M,\n",
    "    'cv2': cv2,\n",
    "    'diffWords': diffWords,\n",
    "    'means_nonzero': means_nozero,\n",
    "    'n_expressed': n_expressed_genes,\n",
    "    'n_genes': n_genes,\n",
    "    'frac_of': frac_of,\n",
    "    'cell_zeros': cell_zeros\n",
    "}\n",
    "\n",
    "with open(\"data_all.pkl\",\"wb\") as file:\n",
    "    pickle.dump(data, file, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load already done analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = load_all_data(\"mca\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correlations with Tabula Muris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import pearsonr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_organs = [\"Bladder\", \"Kidney\", \"Liver\", \"Lung\", \"Pancreas\", \"Spleen\"]\n",
    "\n",
    "organ = common_organs[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tm = load_all_data(\"tm\")[\"freq\"]\n",
    "mca = load_all_data(\"mca\")[\"freq\"]\n",
    "organ=\"all\"\n",
    "\n",
    "#tm = load_tissue(f\"{organ}-counts.csv\", data_source=\"tm\")[\"freq\"]\n",
    "#mca = load_tissue(f\"{organ}\", data_source=\"mca\")[\"freq\"]\n",
    "\n",
    "tm = tm[tm.index.isin(mca.index)]\n",
    "mca = mca[mca.index.isin(tm.index)]\n",
    "tm = pd.Series(name='tm',index=tm.sort_values(ascending=False).index, data=range(1,1+len(tm)))\n",
    "mca = pd.Series(name='mca',index=mca.sort_values(ascending=False).index, data=range(1, len(mca)+1))\n",
    "\n",
    "mca = mca.reindex_like(tm).dropna()\n",
    "tm = tm.reindex_like(mca).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "ax.scatter(mca.values,tm.values, c=\"gray\", s=225, alpha=0.5)\n",
    "\n",
    "bin_avg, bin_edges, _ = stats.binned_statistic(mca.values, tm.values,\"mean\", bins=20)\n",
    "ax.plot(np.linspace(0,1.9e4), np.linspace(0,1.9e4), lw=15, ls=\"--\", label=\"diagonal\")\n",
    "ax.hlines(bin_avg, bin_edges[:-1], bin_edges[1:], lw=15, ls=\"-\", color=\"orange\", label=\"binned average\")\n",
    "\n",
    "ax.set_title(f\"{organ} (Pearson = %.2f)\"%(pearsonr(mca.values, tm.values)[0]), fontsize=35)\n",
    "ax.set_xlabel(\"Mouse Cell Atlas' Rank\", fontsize=35)\n",
    "ax.set_ylabel(\"Tabula Muris' Rank\", fontsize=35)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.legend(loc=\"lower right\", fontsize=35)\n",
    "fig.savefig(f\"rank_correlation_{organ}.pdf\")\n",
    "fig.savefig(f\"rank_correlation_{organ}.eps\")\n",
    "\n",
    "fig.savefig(\"rank_correlation.pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fraction of trascriptome described"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresholded = {}\n",
    "\n",
    "frac_of = data[\"frac_of\"]\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(18,15))\n",
    "for tissue in frac_of.keys():\n",
    "    #if \"Lactation\" in tissue:\n",
    "    #    continue\n",
    "    A_tissue = (frac_of[tissue]['data'] / frac_of[tissue]['N']).sort_values(ascending=False)\n",
    "    A_tissue_sum = np.cumsum(A_tissue.values) / A_tissue.sum() #h thr\n",
    "    ax.plot(A_tissue_sum, label = tissue, c=get_color(tissue), lw=15, alpha=.7)\n",
    "    #if len(A_tissue_sum) < 2:\n",
    "    #    continue\n",
    "    thresholded[tissue] = A_tissue_sum[99]\n",
    "    #thresholded[tissue] = np.argwhere(A_tissue_sum>=0.4).ravel()[0] #v thr\n",
    "ax.set_xscale('log')\n",
    "ax.set_xlim(1, 3e4)\n",
    "ax.set_xlabel(\"Number of genes\", fontsize=35)\n",
    "ax.set_ylabel(\"Fraction of trascriptome described\", fontsize=35)\n",
    "\n",
    "# Shrink current axis by 20%\n",
    "box = ax.get_position()\n",
    "#ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "\n",
    "\n",
    "# Put a legend to the right of the current axis\n",
    "#ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "plt.show()\n",
    "fig.savefig(\"fracof_mca.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All Tissues"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zipf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "for tissue in data[\"frac_of\"].keys():\n",
    "    try:\n",
    "        info = load_tissue(tissue, data_source=data_source)\n",
    "    except:\n",
    "        print(*sys.exc_info())\n",
    "        continue \n",
    "    if \"Lactation\" in tissue:\n",
    "        continue\n",
    "    f = info[\"freq\"]\n",
    "    ax.plot(f.sort_values(ascending=False).values, label=tissue, c=get_color(tissue), lw=15, alpha=.7)\n",
    "    \n",
    "x =np.linspace(1,6e4,10)\n",
    "\n",
    "exp = 0.8\n",
    "ax.plot(x, 1e-1*np.power(x, -0.8), lw=20, c='gray', ls='--')\n",
    "\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "\n",
    "ax.annotate('$k*i^{-%.1f}$'%exp, (1e2,1e-2), fontsize=35)\n",
    "\n",
    "# Shrink current axis by 20%\n",
    "box = ax.get_position()\n",
    "#ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "# Put a legend to the right of the current axis\n",
    "#ax.legend(loc='center left', bbox_to_anchor=(1, 0.5), fontsize=18, ncol=2)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "\n",
    "ax.set_xlim(1,1e5)\n",
    "ax.set_ylim(1e-9,1e0)\n",
    "ax.set_xlabel(\"Rank, $i$\", fontsize=35)\n",
    "ax.set_ylabel(\"Frequency, $f_i$\", fontsize=35)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"zipf_alltissue_{data_source}.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loglog = True\n",
    "filter_mammaryLactation = True\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "for tissue in data[\"frac_of\"].keys():\n",
    "    try:\n",
    "        info = load_tissue(tissue, data_source=data_source)\n",
    "    except:\n",
    "        print(*sys.exc_info())\n",
    "        continue\n",
    "    if (\"Lactation\" in tissue) and filter_mammaryLactation:\n",
    "        continue\n",
    "    M = info[\"M\"]\n",
    "    diff= info[\"diffWords\"]\n",
    "    skip_bins = 15\n",
    "    n_bins=35\n",
    "    if len(M) <= n_bins:\n",
    "        continue\n",
    "    bin_means, bin_edges, binnumber = stats.binned_statistic(M, diff,statistic='mean', bins=np.linspace(M.min(),max(M), n_bins))\n",
    "    bin_stds, _, _ = stats.binned_statistic(M, diff,statistic='std', bins=np.linspace(M.min(),np.quantile(M, 0.9), n_bins))\n",
    "    x = ((bin_edges[:-1]+bin_edges[1:])/2.)[:-skip_bins]\n",
    "    y = bin_means[:-skip_bins]\n",
    "    ax.plot(x, y, marker='o', lw=5, ms=25, alpha=0.8, c=get_color(tissue), label=tissue)\n",
    "    #ax.errorbar(x,y , bin_stds[:-skip_bins], fmt='none', ecolor='orange', elinewidth=3)\n",
    "\n",
    "if \"mca\" in data_source:\n",
    "    ax.set_xlabel(\"UMI per cell\", fontsize=35)\n",
    "else:\n",
    "    ax.set_xlabel(\"Reads per cell\", fontsize=35)\n",
    "ax.set_ylabel(\"Number of expressed genes\", fontsize=35)\n",
    "\n",
    "if loglog:\n",
    "    ax.set_xscale('log')\n",
    "    ax.set_yscale('log')\n",
    "ax.set_xlim(450,max(data[\"M\"])*0.7) #min UMI is 500\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "\n",
    "\n",
    "    \n",
    "# Shrink current axis by 20%\n",
    "box = ax.get_position()\n",
    "#ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "# Put a legend to the right of the current axis\n",
    "#ax.legend(loc='center left', bbox_to_anchor=(1, 0.5), fontsize=18, ncol=2)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(\"heaps_alltissues%s_mca%s.pdf\"%(\"\" if filter_mammaryLactation else \"_noLactation\",\"_loglog\" if loglog else \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import curve_fit\n",
    "def heaps_taylor(X, k):\n",
    "    return k * np.power(X, 2)\n",
    "\n",
    "def heaps_free_exponent(X, C, k):\n",
    "    return C * np.power(X, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loglog = True\n",
    "filter_mammaryLactation = True\n",
    "\n",
    "exps = []\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "for tissue in data[\"frac_of\"].keys():\n",
    "#for tissue in common_organs:\n",
    "    try:\n",
    "        info = load_tissue(tissue)\n",
    "    except:\n",
    "        continue\n",
    "    if (\"Lactation\" in tissue) and filter_mammaryLactation:\n",
    "        continue\n",
    "    #data = load_all_data(\"mca\")\n",
    "    M = info[\"M\"]\n",
    "    diff= info[\"diffWords\"]\n",
    "    n_bins=25\n",
    "\n",
    "    bins = np.logspace(np.log10(M.min()),np.log10(M.max()), n_bins)\n",
    "    #bins = np.linspace(M.min(),max(M), n_bins)\n",
    "    #bins = np.quantile(M,np.linspace(0,1, n_bins))\n",
    "\n",
    "\n",
    "\n",
    "    bin_means, bin_edges, binnumber = stats.binned_statistic(M, diff,statistic='mean', bins=bins)\n",
    "    bin_stds, _, _ = stats.binned_statistic(M, diff,statistic='std', bins=bins)\n",
    "    bin_counts, _, _ = stats.binned_statistic(M, diff,statistic='count', bins=bins)\n",
    "\n",
    "    bin_stds = np.power(bin_stds,2)\n",
    "\n",
    "    mask = bin_counts > 1e2\n",
    "    \n",
    "    x = bin_means[mask]\n",
    "    y = bin_stds[mask]\n",
    "    popt = [1]\n",
    "    try:\n",
    "        popt, pcov = curve_fit(heaps_free_exponent, x,y)\n",
    "        print(tissue, *popt)\n",
    "        exps.append(popt[1])\n",
    "    except:\n",
    "        print(*sys.exc_info())\n",
    "        print(tissue)\n",
    "    finally:\n",
    "        ax.plot(x, y, marker='o', lw=5, ms=25, alpha=0.6, c=get_color(tissue))\n",
    "\n",
    "x = np.linspace(100,5e3)\n",
    "ax.plot(x,x, lw=15,ls=\"--\", c=\"blue\", label=\"Poisson\")\n",
    "ax.plot(x,x*x, lw =15, ls=\"--\", c=\"red\", label=\"Taylor\")\n",
    "ax.set_xlabel(\"Mean number of expressed genes\", fontsize=35)\n",
    "ax.set_ylabel(\"Variance number of expressed genes\", fontsize=35)\n",
    "\n",
    "ax.set_xlim(200,5e3) #min UMI is 500\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "\n",
    "# Shrink current axis by 20%\n",
    "box = ax.get_position()\n",
    "#ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "# Put a legend to the right of the current axis\n",
    "ax.legend(fontsize=35, ncol=1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig(\"heaps_alltissues%s_mca%s_fluctuaction.pdf\"%(\"\" if filter_mammaryLactation else \"_noLactation\",\"_loglog\" if loglog else \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(figsize=(18,15))\n",
    "\n",
    "n_bins = 10\n",
    "bins = np.linspace(0-5/2/n_bins,5+5/2/n_bins,n_bins+2)\n",
    "ax.hist(exps, histtype=\"step\", density=True, color=\"gray\", lw=15, bins=bins)\n",
    "\n",
    "ax.set_xlabel(\"Fit's exponent\", fontsize=35)\n",
    "ax.set_ylabel(\"pdf\", fontsize=35)\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loglog = True\n",
    "filter_mammaryLactation = True\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "info = load_all_data(\"mca\")\n",
    "M = info[\"M\"]\n",
    "diff= info[\"diffWords\"]\n",
    "n_bins=35\n",
    "\n",
    "bins = np.logspace(np.log10(M.min()),np.log10(M.max()), n_bins)\n",
    "#bins = np.linspace(M.min(),max(M), n_bins)\n",
    "#bins = np.quantile(M,np.linspace(0,1, n_bins))\n",
    "\n",
    "bin_means, bin_edges, binnumber = stats.binned_statistic(M, diff,statistic='median', bins=bins)\n",
    "bin_stds, _, _ = stats.binned_statistic(M, diff,statistic='std', bins=bins)\n",
    "bin_counts, _, _ = stats.binned_statistic(M, diff,statistic='count', bins=bins)\n",
    "\n",
    "strong_mask = bin_counts > 5e3\n",
    "mask = bin_counts > 1e2\n",
    "\n",
    "bin_stds = np.power(bin_stds,2)\n",
    "\n",
    "x = bin_means\n",
    "y = bin_stds\n",
    "\n",
    "heapsvar = lambda x,C: C*np.power(x,2)\n",
    "popt, pcov = curve_fit(heapsvar, x[strong_mask], y[strong_mask])\n",
    "print(*popt)\n",
    "print(pcov)\n",
    "\n",
    "ax.plot(x[strong_mask], y[strong_mask], marker='o', lw=15, ms=35, alpha=0.6, c=\"gray\")\n",
    "ax.plot(x[strong_mask], heapsvar(x[strong_mask], *popt), lw=8, ms=25, alpha=0.8, c=\"red\", label=\"Taylor times constant fit\")\n",
    "\n",
    "\n",
    "ax.plot(x,x, lw=15,ls=\"--\", c=\"blue\", label=\"Poisson\")\n",
    "ax.plot(x,x*x, lw =15, ls=\"--\", c=\"red\", label=\"Taylor\")\n",
    "ax.set_xlabel(\"Mean number of expressed genes\", fontsize=35)\n",
    "ax.set_ylabel(\"Variance number of expressed genes\", fontsize=35)\n",
    "\n",
    "ax.set_xlim(300,2e3) #min UMI is 500\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=25)\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "\n",
    "# Shrink current axis by 20%\n",
    "box = ax.get_position()\n",
    "#ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "# Put a legend to the right of the current axis\n",
    "ax.legend(fontsize=35, ncol=1)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig(\"heaps_alltissues%s_mca%s_fluctuaction_all.pdf\"%(\"\" if filter_mammaryLactation else \"_noLactation\",\"_loglog\" if loglog else \"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scaling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CV2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter data\n",
    "var = data[\"var\"].dropna()\n",
    "means = data[\"means\"].dropna()\n",
    "O = data[\"O\"].dropna()\n",
    "O = O[~O.duplicated()]\n",
    "var = var[~var.duplicated()]\n",
    "means = means[var[var>1e-7].index].dropna()\n",
    "var = var.reindex_like(means)\n",
    "O = O.reindex_like(means)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the following cell to select only coding or non-coding genes in GTEx.\n",
    "**(C)** and **(D)** figures of the panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert data_source == \"gtex\", \"Use this piece or code only for GTEx\"\n",
    "genes=pd.read_csv(\"genes_gtex.txt\", header=0, index_col=0)\n",
    "df_genes.head()\n",
    "nc=df_genes[df_genes[\"type_of_gene\"]!=\"protein-coding\"].index\n",
    "pc=df_genes[df_genes[\"type_of_gene\"]==\"protein-coding\"].index\n",
    "genelist = pd.read_csv(\"https://stephenslab.github.io/count-clustering/project/utilities/gene_names_all_gtex.txt\", header=None).values.ravel()\n",
    "pc=pc[pc.isin(genelist)]\n",
    "\n",
    "#select protein coding or noncoding\n",
    "means = means[means.index.isin(pc)]\n",
    "var = var.reindex_like(means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "ax.set_xlim(1e-3, 5e2)\n",
    "\n",
    "#in GTex poisson_limit is 10\n",
    "plotcv2mean(means.values.ravel(), var.values.ravel(), ax=ax, normalisation_str='counts', poisson_limit=1, alpha=0.5, s=255, colorbar=True)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(axis=\"both\", which=\"minor\", width=5, length=10)\n",
    "\n",
    "upbound = len(data[\"M\"])-1\n",
    "x = np.logspace(-5,7)\n",
    "ax.plot(x,[upbound for _ in x], ls=\"--\", lw=15, c=\"orange\", label=\"R-1\")\n",
    "\n",
    "ax.legend(fontsize=30)\n",
    "ax.set_ylim(5e-1,5e5)\n",
    "plt.show()\n",
    "fig.savefig(f\"cv2_mean_dense_{data_source}.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zipf of 100 single cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"mca/mainTable_Bone-Marrow_c-kit.csv\", header=0, index_col=0)\n",
    "M = df.sum(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top100 = M.sort_values(ascending=False).index[:100].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "df[df.columns[df.columns.isin(top100)]].apply(lambda x: ax.plot((x.sort_values(ascending=False)/x.sum()).values), axis=0)\n",
    "\n",
    "x =np.linspace(1,6e3,10)\n",
    "\n",
    "exp = 0.8\n",
    "ax.plot(x, 1e-1*np.power(x, -0.8), lw=20, c='gray', ls='--')\n",
    "ax.annotate('$k*i^{-%.1f}$'%exp, (1e2,1e-2), fontsize=35)\n",
    "\n",
    "\n",
    "ax.set_ylabel(\"$Frequency, f_i$\", fontsize=35)\n",
    "ax.set_xlabel(\"$RANK, i$\", fontsize=35)\n",
    "\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", width=5, length=15)\n",
    "\n",
    "\n",
    "ax.set_xlim(1,2e4)\n",
    "ax.set_ylim(6e-5,1e-1)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(\"zipf_top100_mca.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single tissue: load data\n",
    "Use the following cells to reload already analysed data for a given tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = \"Muscle\"\n",
    "\n",
    "data = load_tissue(tissue, \"data\")\n",
    "\n",
    "for method in [\"mazzolini\"]:\n",
    "    try:\n",
    "        data[method] = load_tissue(tissue, method)\n",
    "\n",
    "        means_null = data[method]['means']\n",
    "        var_null = data[method]['var']\n",
    "        f_null = data[method]['freq']\n",
    "        f_null[np.isnan(f_null)]=0\n",
    "        O_null = data[method]['O']\n",
    "        M_null = data[method][\"M\"]\n",
    "        cv2_null = data[method]['cv2']\n",
    "        diffWords_null = data[method]['diffWords']\n",
    "        means_nozero_null = data[method]['means_nonzero']\n",
    "    except:\n",
    "        print(*sys.exc_info())\n",
    "    \n",
    "\n",
    "means = data['means']\n",
    "var = data['var']\n",
    "f = data['freq']\n",
    "O = data['O']\n",
    "M = data[\"M\"]\n",
    "cv2 = data['cv2']\n",
    "diffWords = data['diffWords']\n",
    "means_nozero = data['means_nonzero']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(f\"mca/mainTable_{tissue}.csv\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = df.sum(axis=0)\n",
    "M=M.sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## U"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def p(o, M, gamma, N):\n",
    "    i = np.arange(1, N+1, step=1)\n",
    "    alpha_i = i**(-gamma)\n",
    "    alpha = np.sum(alpha_i, axis=0)\n",
    "    p_num = (1-o)**(1/M-1)\n",
    "    K = (1-(1-o)**(1/M))**(1+1/gamma)\n",
    "    p_den = gamma*M*N*(alpha**(1/gamma))*K\n",
    "    return p_num/p_den"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbins = 20\n",
    "rang = (0-0.5/nbins, 1+0.5/nbins)\n",
    "\n",
    "#bins = np.logspace(np.log10(1e-2),np.log10(rang[1]),num=nbins)\n",
    "bins = np.linspace(rang[0], rang[1], num=nbins)\n",
    "\n",
    "fig, ax =plt.subplots(figsize=(18,15))\n",
    "ax.hist(np.array(data[\"O\"], dtype=float), histtype='step', lw=15, color='gray', density=True, bins=bins)\n",
    "try:\n",
    "    ax.hist(np.array(data[\"mazzolini\"][\"O\"], dtype=float), histtype='step', lw=15, bins=bins, color='red', ls='--', density=True)\n",
    "except:\n",
    "    pass\n",
    "ax.set_xlabel(\"Occurrence, $o_i$\", fontsize=35)\n",
    "ax.set_ylabel(\"Probability Density Function\", fontsize=35)\n",
    "\n",
    "x_bins=np.arange(1e-2,rang[1],step=(rang[1]-rang[0])/nbins)\n",
    "ax.plot(x_bins,[p(x, M.mean(), 0.8, len(data[\"O\"])) for x in x_bins], lw=15, ls=\"--\", c=\"orange\")\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=15)\n",
    "\n",
    "ax.set_xlim(9e-3,1.1)\n",
    "\n",
    "ax.set_yscale('log')\n",
    "#ax.set_xscale('log')\n",
    "plt.show()\n",
    "fig.savefig(f\"U_{tissue}.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Core Size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "quantiles = np.quantile(M.values, q=np.linspace(0,1,11))[:-1]\n",
    "#idxs = np.linspace(500,M.max(),num=20)\n",
    "\n",
    "sizes = {}\n",
    "sizes_pred = {}\n",
    "thetas_c = [0.6,0.65,0.7,0.75,0.8]\n",
    "for theta in thetas_c:\n",
    "    sizes[\"%.2f\"%theta] = []\n",
    "    sizes_pred[\"%.2f\"%theta] = []\n",
    "\n",
    "P = df.divide(df.sum(0),1).mean(1)\n",
    "\n",
    "#for (m_, m) in zip(M[idxs].index[:-1], M[idxs].index[1:]):\n",
    "for (m_,m) in zip(quantiles[:-1], quantiles[1:]):\n",
    "    print(f\"[{m_},{m})\")\n",
    "    M_avg = (m_+m)/2.\n",
    "    r = len(M[(m_<=M)&(M<m)])\n",
    "    O = df.reindex(columns=M[(m_<=M)&(M<m)].index).dropna(how=\"any\", axis=1).apply(lambda g: len(g[g>0])/float(len(g)), axis=1)\n",
    "    O_pred = np.array(list(map(lambda p: 1-1./r*np.sum(np.exp(-p*M[(m_<=M)&(M<m)].values)),P.values)))\n",
    "    for theta in thetas_c:\n",
    "        sizes[\"%.2f\"%theta].append(len(O[O>theta])/len(O))\n",
    "        sizes_pred[\"%.2f\"%theta].append(len(O_pred[O_pred>theta])/len(O_pred))\n",
    "    del O"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "q_means = (quantiles[1:]+quantiles[:-1])/2.\n",
    "#q_means = (M[idxs].index[1:]+M[idxs].index[:-1])/2.\n",
    "\n",
    "M_bin=q_means\n",
    "\n",
    "N=df.shape[0]\n",
    "\n",
    "def fit_teo(X_data, gamma, thetac, N=df.dropna(how=\"all\", axis=1).shape[0]):\n",
    "    \"\"\"\n",
    "    - gamma: Zipf exponent\n",
    "    - theta_c: Occurrences thresholds\n",
    "    \"\"\"\n",
    "    M_bin = X_data\n",
    "    \n",
    "    i = np.arange(1,N+1, step=1)\n",
    "    alpha_i=i**(-gamma)\n",
    "    alpha=np.sum(alpha_i, axis=0)\n",
    "    # predicted core size \n",
    "\n",
    "    k=M_bin**(1/gamma)/(alpha**(1/gamma)*N)\n",
    "    c=k*(-np.log(1-thetac))**(-1/gamma)\n",
    "    return c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import curve_fit\n",
    "fig,ax = plt.subplots(figsize=(18,15))\n",
    "\n",
    "for (threshold,sizes_arr),(_, sizes_pred_arr), color in zip(sizes.items(),sizes_pred.items(),[\"blue\", \"gray\", \"red\", \"orange\", \"green\"]):\n",
    "    ax.plot(q_means, sizes_arr, lw=10, marker=\"o\", ms=45, c=color, label=f\"data\")\n",
    "    popt = [0.8, float(threshold)]\n",
    "    ax.plot(q_means, sizes_pred_arr, c=\"dark\"+color, lw=15, ls=\"--\", label=\"sampling ($\\\\theta_c$=%.2f)\"%(popt[1]), alpha=0.7)\n",
    "    #ax.plot(q_means, fit_teo(q_means, *popt), c=\"dark\"+color, lw=15, ls=\"--\", label=\"predicted ($\\\\theta_c$=%.2f)\"%(popt[1]), alpha=0.7)\n",
    "\n",
    "threshold = 0.9\n",
    "#ax.plot(q_means, sizes[\"0.9\"], lw=15, marker=\"o\", ms=30, c=\"gray\", label=f\"data with thr={threshold}\")\n",
    "\n",
    "\n",
    "#popt, pcov = curve_fit(fit_teo, q_means, sizes[\"0.9\"], p0=[0.8, 0.95])\n",
    "popt = [0.8, 0.9]\n",
    "#ax.plot(q_means, fit_teo(q_means, *popt), lw=15, ls=\"--\", label=\"predicted (gamma=%.2f; thetac=%.2f)\"%(popt[0],popt[1]), alpha=0.8, c=\"orange\")\n",
    "\n",
    "ax.set_ylim(0,max(list(map(max,list(zip(*sizes.items()))[1])))*1.1)\n",
    "\n",
    "ax.set_ylabel(\"Core size\", fontsize=45)\n",
    "ax.set_xlabel(\"Total reads per cell, $M$\", fontsize=45)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=15)\n",
    "\n",
    "ax.legend(ncol=2,fontsize=30)\n",
    "plt.show()\n",
    "fig.savefig(f\"U_core_prediction_{tissue}_allthr.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[M.index]\n",
    "A = df.sum(1)\n",
    "f = A.sort_values(ascending=False)\n",
    "#df=df[df.index.isin(f[49:4500].index)]\n",
    "#save_model(df,tissue=\"bonemarrow_M100\",name=\"data\", n_bins=15)\n",
    "#mazzolini(np.repeat(M.mean(), len(M)), A/A.sum(), \"bonemarrow_M_avg\", n_bins=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(figsize=(18,15))\n",
    "x = np.arange(1, len(f))\n",
    "ax.set_ylabel(\"$Frequency, f_i$\", fontsize=35)\n",
    "ax.set_xlabel(\"$RANK, i$\", fontsize=35)\n",
    "ax.set_yscale('log')\n",
    "ax.set_xscale('log')\n",
    "#plt.ylim(1e-7,1)\n",
    "ax.plot(x, np.power(x,-0.8)*1e-1, 'g--', lw=10, label='$k*i^{-0.8}$')\n",
    "ax.plot(np.sort(f.dropna())[::-1]/f.sum(), c='blue', lw=15, label='Genes')\n",
    "if \"f_null\" in vars().keys():\n",
    "    ax.plot(np.sort(f_null/f_null.sum())[::-1], c='orange', ls='--', lw=15, label='null_model')\n",
    "    \n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=15)\n",
    "\n",
    "ax.legend(fontsize=35)\n",
    "plt.show()\n",
    "fig.savefig(\"zipf.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predict O"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "P = data[\"freq\"].values\n",
    "M = data[\"M\"].values\n",
    "O = data[\"O\"].values\n",
    "R = M.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "O_real = np.array([])\n",
    "O_pred = np.array([])\n",
    "O_pred_ds = np.array([])\n",
    "\n",
    "\n",
    "step = 1500\n",
    "for start in range(0,len(O),step):\n",
    "    print(start)\n",
    "    mask = np.repeat(False, len(O))\n",
    "    mask[start:start+step]=True\n",
    "\n",
    "    O_real = np.concatenate([O_real,O[mask]])\n",
    "    O_pred = np.concatenate([O_pred,list(map(lambda p: 1-1/R*np.sum(np.exp(-p*M)),P[mask]))])\n",
    "    \n",
    "genes = data[\"freq\"].index[O_real < (0.8*O_pred)].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = data[\"freq\"]\n",
    "\n",
    "pos_mask = (~f.isna()) & (f>0)\n",
    "\n",
    "g_high = f.index[O_pred > O_real + 0.2].values\n",
    "g_low = f.index[O_pred < O_real * 0.6 - 0.03].values\n",
    "\n",
    "mask_high = (f.index.isin(g_high)) & (pos_mask)\n",
    "mask_low = (f.index.isin(g_low)) & (pos_mask)\n",
    "#mask_nc = (f.index.isin(nc)) & (pos_mask)\n",
    "\n",
    "f_sorted = f[f>0].sort_values(ascending=False)\n",
    "mask_sorted_high = f_sorted.index.isin(g_high)\n",
    "mask_sorted_low = f_sorted.index.isin(g_low)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "ax.scatter(O_real, O_pred, s=200, c='gray')\n",
    "ax.plot([0,1], [0,1], color = \"black\", ls='--', lw=15, alpha=0.9)\n",
    "\n",
    "ax.scatter(O_real[mask_high], O_pred[mask_high], s=200, c='b')\n",
    "ax.scatter(O_real[mask_low], O_pred[mask_low], s=200, c='red')\n",
    "\n",
    "#ax.scatter(O_real[mask_nc], O_pred[mask_nc], s=200, c='red')\n",
    "\n",
    "\n",
    "ax.set_xlabel(\"$O_i$ from data\", fontsize=35)\n",
    "ax.set_ylabel(\"$O_i$ predicted\", fontsize=35)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", width=5, length=10)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig(f\"O_real-O_pred_{tissue}.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "\n",
    "nbins=30\n",
    "maxbin = np.sqrt(2)/2\n",
    "minbin = (O_pred-O_real).min()\n",
    "bins = np.arange(minbin, maxbin, step=(maxbin-minbin)/nbins)\n",
    "\n",
    "ax.hist(O_pred-O_real,  color='gray', bins=20)\n",
    "\n",
    "ax.set_xlabel(\"$O_i$ predicted - $O_i$ from data\", fontsize=35)\n",
    "ax.set_ylabel(\"pdf\", fontsize=35)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", width=5, length=10)\n",
    "\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xlim(-0.1, 0.1+1./np.sqrt(2))\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig(f\"O_real-O_pred_{tissue}_hist.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_findzeros = pd.DataFrame().join(pd.Series(name=\"mean\", data=data[\"means\"]), how=\"outer\").join(pd.Series(name=\"var\", data=data[\"var\"]), how=\"inner\")\n",
    "df_findzeros[\"cv2\"] = df_findzeros[\"var\"] / df_findzeros[\"mean\"] / df_findzeros[\"mean\"]\n",
    "df_findzeros[\"poisson_dist\"] = (df_findzeros[\"cv2\"] * df_findzeros[\"mean\"])\n",
    "\n",
    "df_findzeros = df_findzeros.join(pd.Series(name=\"O\", index = data[\"O\"].index, data=data[\"O\"]), how=\"outer\")\n",
    "df_findzeros = df_findzeros.join(pd.Series(name=\"O_real\", index = data[\"O\"].index, data=O_real), how=\"outer\")\n",
    "df_findzeros = df_findzeros.join(pd.Series(name=\"O_pred\", index = data[\"O\"].index, data=O_pred), how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(18,15))\n",
    "\n",
    "ax.scatter(df_findzeros[\"poisson_dist\"],df_findzeros[\"O_pred\"]-df_findzeros[\"O_real\"], s=225, alpha=0.8, c=\"gray\")\n",
    "\n",
    "ax.set_xlabel(\"Fano factor\", fontsize=35)\n",
    "ax.set_ylabel(\"O_pred - O_real\", fontsize=35)\n",
    "ax.tick_params(labelsize=25)\n",
    "\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", width=5, length=10)\n",
    "\n",
    "ax.set_xscale(\"log\")\n",
    "#ax.set_yscale(\"log\")\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"OrealOpred_Fano_{tissue}.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"OrealOpred_overlist_{tissue}_low.txt\",\"w\") as f:\n",
    "    list(map(lambda g: f.write(g+\"\\n\"), filter(lambda g: \"Rik\" not in g, data[\"freq\"].index.values[mask_low])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# theoretical prediction\n",
    "def h(m, gamma: list, N, F1=122, F2=4100):\n",
    "    g1, g2 = gamma\n",
    "    i = np.arange(1, N+1, step=1)\n",
    "    \n",
    "    b = 1./(F1**(g1-g2)*np.sum([_i**(-g1) for _i in i[:F1]])+np.sum([_i**(-g2) for _i in i[F1:F2]]) + (F2**(-g2))*np.exp(0.002*F2)*np.sum([np.exp(-0.002*_i) for _i in i[F2:]]))\n",
    "    c = b * (F2**(-g2))*np.exp(0.002*F2)\n",
    "    a = b * F1**(g1-g2)\n",
    "    return N - np.sum((1-a*np.power(i[:F1], -g1))**m, axis=0) - np.sum((1-b*np.power(i[F1:F2], -g2))**m, axis=0) - np.sum((1-c*np.exp(-0.002*i[F2:]))**m, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(28,24))\n",
    "\n",
    "skip_bins=5\n",
    "ax.scatter(data[\"M\"], data[\"diffWords\"], color=\"gray\", s=200, label=\"data\")\n",
    "\n",
    "ax.scatter(data[\"mazzolini\"][\"M\"], data[\"mazzolini\"][\"diffWords\"], color=\"darkorange\", s=250, label=\"sampling_ensamble\")\n",
    "\n",
    "\n",
    "bin_means, bin_edges, binnumber = stats.binned_statistic(data[\"M\"], data[\"diffWords\"],statistic='mean', bins=np.logspace(np.log10(data[\"M\"].min()),np.log10(max(data[\"M\"])), 25))\n",
    "\n",
    "ax.plot((bin_edges[:-1][:-skip_bins] + bin_edges[1:][:-skip_bins])/2, bin_means[:-skip_bins], color='blue', lw=10, marker=\"o\", ms=25, label='binned average')\n",
    "\n",
    "\n",
    "x_bins=np.logspace(np.log10(data[\"M\"].min()),np.log10(max(data[\"M\"])))\n",
    "ax.plot(x_bins, [h(x, [0.5, 1.0], len(data[\"freq\"].dropna()), F1=122, F2=4100) for x in x_bins], lw=15, alpha=0.6, ls=\"--\", c=\"red\", label=\"theoretical prediction\")\n",
    "\n",
    "ax.set_ylabel(\"Number of genes expressed\", fontsize=35)\n",
    "ax.set_xlabel(\"UMI per cell, $M_s$\", fontsize=35)\n",
    "ax.tick_params(labelsize=35, width=8, length=20)\n",
    "ax.tick_params(which=\"minor\", labelsize=35, width=5, length=15)\n",
    "\n",
    "ax.legend(fontsize=25)\n",
    "\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"heaps_{tissue}_mazzolini_teo.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
